### ISSUE TITLE
Multi-world Support

### ISSUE THREAD

**Issue by creatorfromhell** [lokka30/Treasury#77]
I believe that multi-world support should be at least provided in an optional capacity. The reason for this is the fact that economy plugins won't be able to always know the world required for a transaction. 

Examples

Let's say we have a third-party plugin that charges for changing worlds, but they need to process a transaction for the world the player wishes to go to. The economy plugin will only have access to the world they're in currently, not the one they're attempting to go to.

Let's say a server has resource world, and players get paid money for resources they mine in that world. Obviously they won't want their money to be added into that world's balance information, but again that's the only world the economy plugin will have access to. The only alternative would force a system onto the economy plugin or resource world plugin to group those two balances together, which isn't really ideal.

Another example is if a quest plugin requires a specific amount of currency in another world, how would they check for this? They wouldn't be able to.

**Comment by MrNemo64** [comment]
We discussed this topic here but I'm willing to reopen it

https://github.com/lokka30/Treasury/issues/30

tl;dr
Instead of per world balance a corrency would only be usable in certain world

**Comment by creatorfromhell** [comment]
> We discussed this topic here but I'm willing to reopen it
> 
> #30
> 
> tl;dr Instead of per world balance a corrency would only be usable in certain world

This is again forcing a choice that should be up to the provider itself onto the provider. What about for systems that use global currencies?

**Comment by MrNemo64** [comment]
If you have a currency that you can only use in the _resources_ world that ends there, you may only use it there but you can always  convert it to another currency that you can use globally
A consumer plugin would just need to get the currencies avilable in X world and use those, the provider just has to know in what world currencies are able to be used and where a currency can be used can be left to the server owner to be made as complex as they like.

**Comment by creatorfromhell** [comment]
> If you have a currency that you can only use in the _resources_ world that ends there, you may only use it there but you can always convert it to another currency that you can use globally A consumer plugin would just need to get the currencies avilable in X world and use those, the provider just has to know in what world currencies are able to be used and where a currency can be used can be left to the server owner to be made as complex as they like.

That's the problem with this system is the provider is being told what to know... The provider should be dictating what is happening, otherwise Treasury would be taking on the design choices and here's what'll happen:

Plugin user: Why can't I have global currencies?
Plugin Developer: talk to treasury it's on their end, they made the choice
- tickets flood into Treasury because they're forcing provider's hands.

**Comment by lokka30** [comment]
We might be able to add back per-world balances. I just hope that we can maintain ease of use in the API and not create too much confusion with how to use these methods.
Perhaps `Optional<String>` for a world name in transactions, `Optional.empty()` representing 'global balance'.

**Comment by MrIvanPlays** [comment]
Words create confusion in vault, and I'm fairly sure they would here. 

I'm against adding this into the treasury API. If an economy provider wants to have such feature, best bet is to create a per world currency and make it work only if they are in that specific world.

**Comment by Jikoo** [comment]
My original reply on this was on Spigot:

> On more thought, per-world balances are somewhat useless - if users want per-world balances, they can add currencies that are only used in certain worlds. If all the shops in RewardLand only accept NeatWorldDosh your CoolWorldBucks are useless there despite being visible and available. It offloads the strain of multi-world management from the API/implementation onto the server owner configuring the setup. They can make it as complex or as simple as their needs without any extra effort on the programmers' ends.

In addition to this, all usages have to be written with multiworld support in mind to be usable in a multi-world environment. Multi-currency negates the need for multi-world functionality, especially since it is a load of strain on developers that is then removed to the server administrator.

Going through your examples I'm not sure why they would require Treasury to offer per-world parameters in addition to multiple currencies.

> Let's say we have a third-party plugin that charges for changing worlds, but they need to process a transaction for the world the player wishes to go to. The economy plugin will only have access to the world they're in currently, not the one they're attempting to go to.

Charge them in a currency used primarily in the world they're going to - just because currencies are global does not mean you have to use each one in every single world. Restrict access to the currency and you have created a world-specific currency. This simplifies things on both developer and users' ends - developers don't have to worry about world grouping (world, world_nether, etc.) and destination worlds, users simply have to restrict access to things that use the currency they've decided should only be accessible from the other world. The "group" of worlds is the currency itself.

An aside: this can probably be used for a Vault compat layer, world (group) name <-> currency name

> Let's say a server has resource world, and players get paid money for resources they mine in that world. Obviously they won't want their money to be added into that world's balance information, but again that's the only world the economy plugin will have access to. The only alternative would force a system onto the economy plugin or resource world plugin to group those two balances together, which isn't really ideal.

This is basically the same thing and is solved by setting up currency access.

> Another example is if a quest plugin requires a specific amount of currency in another world, how would they check for this? They wouldn't be able to.

Again, specific currency. Maybe I earn "Quest Unlock" for completing quests, and I need 50 "Quest Unlock" to accept this next quest. If I can only earn "Quest Unlock" by working in the mines in another world, is it not specific to the other world? Globally accessible, but the player can only spend it on questing and earn it by mining in a specific world.

The whole juggling back and forth of "per-world" stuff that's being accessed from other worlds really doesn't make sense. The "world" argument is just a grouping argument that is effectively the same as multiple balances for the same currency with a unique identifier. That gets super confusing to deal with, and most people won't. Think about how few plugins add Vault's multiworld support. Think about how few plugins then implement that in their usages. Even fewer, 1) because it's hassle and 2) because most implementations don't support it. With multiple currencies being a core feature, the entire pain just vanishes.

**Comment by lokka30** [comment]
My opinion is on par with Jikoo's and Ivan's concerning the addition of direct multi-world support in the API.

**Comment by MrIvanPlays** [comment]
since there is not very much interest in multi world support and the counterarguments that we have multiple currencies and that all of the maintainers (lokka30, jikoo and me) are against this I'm closing this issue. if anyone shows interest I'll surely reopen it, but they need to have very strong arguments for adding such given the counterargument of multiple currencies.
